lxd/instances_get.go

```diff
@@ -356,126 +358,126 @@ func doInstancesGet(d *Daemon, r *http.Request) (interface{}, error) {
 		// on this node
 		if isClusterNotification(r) && address != "" {
 			continue
 		}
 
-		// Mark containers on unavailable nodes as down
+		// Mark containers on unavailable projectInstanceToNodeName as down
 		if mustLoadObjects && address == "0.0.0.0" {
-			for _, instanceName := range instanceNames {
+			for _, projectInstance := range projectsInstances {
 				if recursion < 2 {
-					resultListAppend(instanceName, api.Instance{}, fmt.Errorf("unavailable"))
+					resultListAppend(projectInstance, api.Instance{}, fmt.Errorf("unavailable"))
 				} else {
-					resultFullListAppend(instanceName, api.InstanceFull{}, fmt.Errorf("unavailable"))
+					resultFullListAppend(projectInstance, api.InstanceFull{}, fmt.Errorf("unavailable"))
 				}
 			}
 
 			continue
 		}
 
 		// For recursion requests we need to fetch the state of remote
-		// containers from their respective nodes.
+		// containers from their respective projectInstanceToNodeName.
 		if mustLoadObjects && address != "" && !isClusterNotification(r) {
 			wg.Add(1)
-			go func(address string, containers []string) {
+			go func(address string, projectsInstances [][2]string) {
 				defer wg.Done()
 
 				if recursion == 1 {
 					cs, err := doContainersGetFromNode(filteredProjects, address, allProjects, networkCert, d.serverCert(), r, instanceType)
 					if err != nil {
-						for _, name := range containers {
-							resultListAppend(name, api.Instance{}, err)
+						for _, projectInstance := range projectsInstances {
+							resultListAppend(projectInstance, api.Instance{}, err)
 						}
 
 						return
 					}
 
 					for _, c := range cs {
-						resultListAppend(c.Name, c, nil)
+						resultListAppend([2]string{c.Name, c.Project}, c, nil)
 					}
 
 					return
 				}
 
 				cs, err := doContainersFullGetFromNode(filteredProjects, address, allProjects, networkCert, d.serverCert(), r, instanceType)
 				if err != nil {
-					for _, name := range containers {
-						resultFullListAppend(name, api.InstanceFull{}, err)
+					for _, projectInstance := range projectsInstances {
+						resultFullListAppend(projectInstance, api.InstanceFull{}, err)
 					}
 
 					return
 				}
 
 				for _, c := range cs {
-					resultFullListAppend(c.Name, c, nil)
+					resultFullListAppend([2]string{c.Name, c.Project}, c, nil)
 				}
-			}(address, instanceNames)
+			}(address, projectsInstances)
 
 			continue
 		}
 		if !mustLoadObjects {
-			for _, instanceName := range instanceNames {
+			for _, projectInstance := range projectsInstances {
 				instancePath := "instances"
 				if strings.HasPrefix(mux.CurrentRoute(r).GetName(), "container") {
 					instancePath = "containers"
 				} else if strings.HasPrefix(mux.CurrentRoute(r).GetName(), "vm") {
 					instancePath = "virtual-machines"
 				}
-				url := fmt.Sprintf("/%s/%s/%s", version.APIVersion, instancePath, instanceName)
+				url := fmt.Sprintf("/%s/%s/%s", version.APIVersion, instancePath, projectInstance[1])
```

Comment:
@stgraber @markylaing I've spotted another issue, the URL returned here for a multi-project response won't contain the correct `?project=` URL parameter.

@stgraber is there any reason we don't add the `project=` param to all URLs apart from the default project in the non-recursive instance list response?

Reply:
Also we should modify this to use the `api.NewURL()` builder.